src/SmurfBuilder.cpp

```diff
@@ -82,106 +77,178 @@ void SmurfBuilder::FlushStash(){
+    int primary_shape[2] = {num_keys, nsamps};
+    auto primary_ts = G3SuperTimestreamPtr(new G3SuperTimestream());
+    primary_ts->names = G3VectorString();
+    for (auto name : primary_keys)
+        primary_ts->names.push_back(name);
+    
     G3VectorTime sample_times = G3VectorTime(nsamps);
-    for (auto key : primary_keys)
-        primary_vec.push_back(G3VectorIntPtr(new G3VectorInt(nsamps, 0)));
 
-    boost::shared_ptr<G3TimesampleMap> primary_map = boost::make_shared<G3TimesampleMap>();
+    if (debug_)
+        copy_start = std::chrono::system_clock::now();
 
+    // Read data in to G3 Objects
     int sample = 0;
-    for (auto x = read_stash_.begin(); x != read_stash_.end(); x++, sample++){
-        auto hdr = (*x)->sp->getHeader();
-
-        sample_times[sample] = (*x)->GetTime();
-
-        (*primary_vec[0])[sample] = hdr->getUnixTime();
-        (*primary_vec[1])[sample] = hdr->getFluxRampIncrement();
-        (*primary_vec[2])[sample] = hdr->getFluxRampOffset();
-        (*primary_vec[3])[sample] = hdr->getCounter0();
-        (*primary_vec[4])[sample] = hdr->getCounter1();
-        (*primary_vec[5])[sample] = hdr->getCounter2();
-        (*primary_vec[6])[sample] = hdr->getAveragingResetBits();
-        (*primary_vec[7])[sample] = hdr->getFrameCounter();
-        (*primary_vec[8])[sample] = hdr->getTESRelaySetting();
+    for (auto it = start; it != stop; it++, sample++){
+        sample_times[sample] = (*it)->GetTime();
+
+        auto hdr = (*it)->sp->getHeader();
+        primary_buffer[sample + 0 * nsamps] = hdr->getUnixTime();
+        primary_buffer[sample + 1 * nsamps] = hdr->getFluxRampIncrement();
+        primary_buffer[sample + 2 * nsamps] = hdr->getFluxRampOffset();
+        primary_buffer[sample + 3 * nsamps] = hdr->getCounter0();
+        primary_buffer[sample + 4 * nsamps] = hdr->getCounter1();
+        primary_buffer[sample + 5 * nsamps] = hdr->getCounter2();
+        primary_buffer[sample + 6 * nsamps] = hdr->getAveragingResetBits();
+        primary_buffer[sample + 7 * nsamps] = hdr->getFrameCounter();
+        primary_buffer[sample + 8 * nsamps] = hdr->getTESRelaySetting();
+
+        for (int i = 0; i < N_TES_BIAS; i++)
+            tes_buffer[sample + i * nsamps] = hdr->getTESBias(i);
+
+        for (int i = 0; i < nchans; i++){
+            data_buffer[sample + i * nsamps] = (*it)->sp->getData(i);
+        }
     }
 
-    primary_map->times = sample_times;
+    PyGILState_STATE gstate;
+    gstate = PyGILState_Ensure();
 
-    int i = 0;
-    for (auto key : primary_keys){
-        primary_map->insert(std::make_pair(key, primary_vec[i]));
-        i++;
-    }
+    data_ts->times = sample_times;
+    data_ts->SetDataFromBuffer((void*)data_buffer, 2, data_shape, NPY_INT32,
+            std::pair<int,int>(0, nsamps));
+    data_ts->Options(
+        enable_compression_, flac_level_, bz2_work_factor_, data_encode_algo_,
+        time_encode_algo_
+    );
+
+    tes_ts->times = G3VectorTime(sample_times);
```

Comment:
Breaking:  I think I just found a bug where G3VectorTime(G3VectorTime &) does not work properly -- the result has twice as many elements (padded with zeros) as the thing you passed in.  Please check.  Will submit report to spt3g_software; please find a work-around for now.

Reply:
Nevermind ... the bug doesn't affect this constructor in C++.